TT Lab/Rendering/EditingContext.cs

```diff
@@ -0,0 +1,336 @@
+            }
+            if (transformMode == TransformMode.ROTATE)
+            {
+                var k = 0.2f;
+                if (transformAxis == TransformAxis.X)
+                {
+                    RotateX(k * delta);
+                }
+                else if (transformAxis == TransformAxis.Y)
+                {
+                    RotateY(k * delta);
+                }
+                else if (transformAxis == TransformAxis.Z)
+                {
+                    RotateZ(k * delta);
+                }
+            }
+        }
+
+        public void ToggleTranslate()
+        {
+            if (transforming) {
+                return;
+            }
+            if (transformMode != TransformMode.TRANSLATE)
+            {
+                transformMode = TransformMode.TRANSLATE;
+            }
+            else
+            {
+                transformMode = TransformMode.SELECTION;
+            }
+            transformAxis = TransformAxis.NONE;
+        }
+
+        public void ToggleRotate()
+        {
+            if (transforming)
+            {
+                return;
+            }
+            if (transformMode != TransformMode.ROTATE)
+            {
+                transformMode = TransformMode.ROTATE;
+            }
+            else
+            {
+                transformMode = TransformMode.SELECTION;
+            }
+            transformAxis = TransformAxis.NONE;
+        }
+
+        public void SetTransformAxis(TransformAxis axis)
+        {
+            if (transformAxis == axis)
+            {
+                transformAxis = TransformAxis.NONE;
+            } else
+            {
+                transformAxis = axis;
+            }
+            if (transforming)
+            {
+                UpdateTransform(endPos.x, endPos.y);
+            }
+        }
+
+        public void ToggleSpace()
+        {
+            transformSpace = transformSpace == TransformSpace.WORLD ? TransformSpace.LOCAL : TransformSpace.WORLD;
+            if (transforming)
+            {
+                UpdateTransform(endPos.x, endPos.y);
+            }
+        }
+
+        private mat4 startTransform;
+        private vec2 startPos;
+        private vec2 endPos;
+        private bool transforming = false;
+
+        //TODO: must affect InstanceData too
+        private void Translate(vec3 offset)
+        {
+            if (transformSpace == TransformSpace.LOCAL)
+            {
+                selectedInstance.GetRenderable().LocalTransform = startTransform * mat4.Translate(offset);
+            } 
+            else
+            {
+                selectedInstance.GetRenderable().LocalTransform = mat4.Translate(offset) * startTransform;
+            }
+        }
+
+        private void RotateX(float value)
+        {
+            if (transformSpace == TransformSpace.LOCAL)
+            {
+                selectedInstance.GetRenderable().LocalTransform = startTransform * mat4.RotateX(value * 3.14f / 180.0f);
```

Comment:
Since GetRenderable is used so often let's cache it when selectedInstance is updated